proxy/src/proxy_metrics.rs

```diff
@@ -0,0 +1,188 @@
+//!
+//! Periodically collect proxy consumption metrics
+//! and push them to a HTTP endpoint.
+//!
+//!
+use std::{collections::HashMap, time::Duration};
+
+use tracing::{debug, error, log::info, trace};
+
+use consumption_metrics::{
+    ConsumptionMetricKind, ProxyConsumptionMetric, ProxyConsumptionMetricsKey, ProxyEventChunk,
+    CHUNK_SIZE,
+};
+
+use chrono::{DateTime, Utc};
+
+pub async fn collect_metrics(
+    metric_collection_endpoint: &reqwest::Url,
+    metric_collection_interval: Duration,
+    hostname: String,
+) -> anyhow::Result<()> {
+    scopeguard::defer! {
+        info!("collect_metrics has shut down");
+    }
+
+    let mut ticker = tokio::time::interval(metric_collection_interval);
+
+    info!(
+        "starting collect_metrics. metric_collection_endpoint: {}",
+        metric_collection_endpoint
+    );
+
+    // define client here to reuse it for all requests
+    let client = reqwest::Client::new();
+    let mut cached_metrics: HashMap<ProxyConsumptionMetricsKey, (u64, DateTime<Utc>)> =
+        HashMap::new();
+
+    loop {
+        tokio::select! {
+            _ = ticker.tick() => {
+
+                match collect_metrics_task(&client, &mut cached_metrics, metric_collection_endpoint, hostname.clone()).await
+                {
+                    Err(e) => {
+                        error!("Failed to send consumption metrics: {} ", e);
+                    },
+                    Ok(_) => { trace!("collect_metrics_task completed successfully") },
+                }
+            }
+        }
+    }
+}
+
+pub fn gather_proxy_io_bytes_per_client() -> Vec<(ProxyConsumptionMetricsKey, (u64, DateTime<Utc>))>
+{
+    let mut current_metrics: Vec<(ProxyConsumptionMetricsKey, (u64, DateTime<Utc>))> = Vec::new();
+    let metrics = prometheus::default_registry().gather();
+
+    for m in metrics {
+        if m.get_name() == "proxy_io_bytes_per_client" {
+            for ms in m.get_metric() {
+                let direction = ms
+                    .get_label()
+                    .iter()
+                    .find(|l| l.get_name() == "direction")
+                    .unwrap()
+                    .get_value();
+
+                let endpoint_id = ms
+                    .get_label()
+                    .iter()
+                    .find(|l| l.get_name() == "endpoint_id")
+                    .unwrap()
+                    .get_value();
```

Comment:
It's possible to do this in one traversal using `for` and two options.

Reply:
there are not too many labels, so I'd keep it as is.
to me that looks more readable than alternative